python/pyarrow/dataset.py

```diff
@@ -788,7 +805,8 @@ def file_visitor(written_file):
     if max_partitions is None:
         max_partitions = 1024
 
-    partitioning = _ensure_write_partitioning(partitioning)
+    partitioning = _ensure_write_partitioning(partitioning, schema=schema,
```

Comment:
It seems that schema can be None at this point, in which case `_ensure_write_partitioning` might error. It seems that when passing a Scanner `schema` must be None (or when constructing a Scanner from an iterable, the `schema` is set to None after constructing the scanner, see L781 above). But in those cases, we can get the `schema` from the scanner to pass here.

Reply:
Should have addressed those cases and added tests for them